### PR TITLE
feat: add standard fast confirmation metrics

### LINKED ISSUE #9545: Implement standard FCR metrics
- https://github.com/ethereum/beacon-metrics/pull/22

### PR DESCRIPTION
**Pull request by nflaig**
**Motivation**
  
Implement the standard fast confirmation metrics defined in https://github.com/ethereum/beacon-metrics/pull/22
  
**Description**

Adds the four standard FCR metrics alongside the existing non-standard `lodestar_*` metrics (kept as-is, even where they overlap):
  - `beacon_fast_confirmation_slot` — slot of the most recent confirmed block
  - `beacon_fast_confirmation_reorgs_total` — confirmed block no longer an ancestor of head
  - `beacon_fast_confirmation_fallbacks_total` — revert of the confirmed block to finalized
  - `beacon_fast_confirmation_restarts_total` — restart from the observed unrealized justified checkpoint

The per-rule `reason` is overwritten by later rules in the pipeline (e.g. a restart is followed by advancing to the latest confirmed descendant), so `runFastConfirmationRules` now accumulates the set of reasons decided across all rules and reports `didReorg` / `didRestart` from it.

Closes #9545

### CODE REVIEW

**Comment by nazarhussain** [comment]
@chatgpt-codex-connector  @gemini-code-assist  Re-review the changes.

**Comment by nazarhussain** [comment]
@lodekeeper Please review this PR.

**Review by lodekeeper — APPROVED** [review]
Reviewed at `52a660b7`. Both earlier bot findings are addressed at head, and I traced the logic through the rule pipeline:

- **fallbacks** (gemini HIGH): now gated on `decision.didReset && decision.confirmedRoot === snapshot.finalizedRoot`, so a reset that later advances forward (restart / confirmed-descendant) no longer counts as a fallback to finality.
- **reorgs** (codex P2): derived from ancestry (`!isAncestor(headRoot, confirmedRoot)`) instead of the reset reason, so it correctly catches the epoch-behind **and** not-ancestor case where `resetIfBehindOrNotAncestorOrUnsafe` records `ResetBehind` (which takes precedence). The new test covers exactly that.

Correctness points I checked and am happy with:
- `isAncestor` is reflexive (`getAncestor(head, headSlot) === head`), so `confirmed === head` doesn't produce a false reorg.
- When the confirmed block is non-canonical, `resetIfBehindOrNotAncestorOrUnsafe` resets it back to finalized in the same run, so `reorgs` is effectively edge-triggered — it won't keep re-counting the same reorg on subsequent slots.
- `didRestart` reads from the `reasons` set rather than the final `decision.reason`, so it survives `advanceToLatestConfirmedDescendant` overwriting the reason.

Two non-blocking notes:

1. **Metric prefix mix** — the four new standardized metrics use `beacon_fast_confirmation_*`, but `confirmed_epoch` (gauge) and `resets` (counter) stay on `lodestar_fast_confirmation_*`. Is that intentional (only the standardized set moves to `beacon_`), or should those migrate too? Just confirming intent, since dashboards/alerts may key off the prefix.
2. **`resets` gauge→counter** — good fix (it was `_total`-named but registered as a gauge). Heads-up that the "Reset Total" dashboard panel still plots it raw; `rate()`/`increase()` reads better for a counter now. Optional.

LGTM otherwise — logic is correct and well-tested.

File: packages/fork-choice/src/forkChoice/fastConfirmation/fastConfirmationRule.ts [hunk 1]
```diff
@@ -144,15 +145,26 @@ export class FastConfirmationRule implements IFastConfirmationRule {
...
     const confirmedBlock = ctx.getBlock(result.confirmedRoot);
     if (confirmedBlock) {
       this.metrics.fastConfirmation.confirmedSlot.set(confirmedBlock.slot);
       this.metrics.fastConfirmation.confirmedEpoch.set(computeEpochAtSlot(confirmedBlock.slot));
+      this.metrics.fastConfirmation.slot.set(confirmedBlock.slot);
     }
     if (result.didReset) {
       this.metrics.fastConfirmation.resets.inc();
+      this.metrics.fastConfirmation.fallbacks.inc();
+    }
```
[the remark this answers is not in the document]
Comment (lodekeeper) [reply packages/fork-choice/src/forkChoice/fastConfirmation/fastConfirmationRule.ts:160]: Valid against the initial commit — and already addressed in 59c53f57f1. `fallbacks` no longer fires on every reset: `runFastConfirmationRules` now sets `didFallback = didReset && confirmedRoot === finalizedRoot`, and `beacon_fast_confirmation_fallbacks_total` only increments on that. So a reset that a later rule advances forward from (e.g. to the observed justified checkpoint) is classified as a restart, not a fallback — exactly your point. Covered by the `advances to observed justified` unit test (asserts `didFallback=false`, `didRestart=true`).

File: packages/fork-choice/src/forkChoice/fastConfirmation/rules.ts [hunk 2]
```diff
@@ -110,15 +111,26 @@ export function runFastConfirmationRules(
...
     decision = rule(snapshot, ctx, store, cache, decision, logger);
+    reasons.add(decision.reason);
   }
-  return decision;
+
+  return {
+    confirmedRoot: decision.confirmedRoot,
+    didReset: decision.didReset,
+    reason: decision.reason,
+    didReorg: reasons.has(FastConfirmationDecisionReason.ResetNotAncestor),
```
[the remark this answers is not in the document]
Comment (lodekeeper) [reply packages/fork-choice/src/forkChoice/fastConfirmation/rules.ts]: Valid against the initial commit — already addressed in 6809c71ab7. `didReorg` is now derived from ancestry rather than the reset reason: `didReorg = initialConfirmedBlock !== null && !isAncestor(head, confirmedRoot)`. So the case you flagged — confirmed block both epoch-behind *and* no longer an ancestor of head, where `resetIfBehindOrNotAncestorOrUnsafe` records `ResetBehind` ahead of `ResetNotAncestor` — still counts as a reorg. New regression test `reports a reorg when the confirmed block is both epoch-behind and not an ancestor of head` asserts `didReorg=true` for exactly that.

**Comment by lodekeeper** [comment]
Reviewed 👍 Both bot findings were valid against the initial commit (4a3d0e5) and are already resolved on the current head:

- **Gemini (fallbacks)** — fixed in 59c53f57f1: `didFallback = didReset && confirmedRoot === finalizedRoot`, so a reset that's advanced forward to the observed justified checkpoint is counted as a restart, not a fallback to finality.
- **Codex (reorgs)** — fixed in 6809c71ab7: `didReorg` now comes from ancestry (`!isAncestor(head, confirmedRoot)`) instead of the reset reason, so the epoch-boundary case where `ResetBehind` takes precedence over `ResetNotAncestor` is still counted.

Both have dedicated regression tests, and the fast-confirmation unit suite passes locally (15/15). Metric names/types line up with beacon-metrics#22, and the `resets` gauge→counter change is correct (it's only ever `.inc()`'d). LGTM.

**Review by wemeetagain — DISMISSED** [review]

File: dashboards/lodestar_fork_choice.json [hunk 3]
```diff
@@ -1700,6 +1700,126 @@
...
+      "pluginVersion": "10.4.1",
+      "targets": [
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "${DS_PROMETHEUS}"
+          },
+          "disableTextWrap": false,
+          "editorMode": "builder",
+          "expr": "beacon_fast_confirmation_reorgs_total",
```
Comment (nflaig) [inline dashboards/lodestar_fork_choice.json]: aren't these counters? we should use `rate()`, please double check

File: dashboards/lodestar_fork_choice.json [hunk 4]
```diff
@@ -1700,6 +1700,126 @@
...
+          "useBackend": false
+        },
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "${DS_PROMETHEUS}"
+          },
+          "disableTextWrap": false,
+          "editorMode": "builder",
+          "expr": "beacon_fast_confirmation_fallbacks_total",
```
Comment (nflaig) [inline dashboards/lodestar_fork_choice.json]: same as above use rate

File: dashboards/lodestar_fork_choice.json [hunk 5]
```diff
@@ -1700,6 +1700,126 @@
...
+          "useBackend": false
+        },
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "${DS_PROMETHEUS}"
+          },
+          "disableTextWrap": false,
+          "editorMode": "builder",
+          "expr": "beacon_fast_confirmation_restarts_total",
```
Comment (nflaig) [inline dashboards/lodestar_fork_choice.json]: another one to use rate

File: packages/fork-choice/src/forkChoice/fastConfirmation/fastConfirmationRule.ts [hunk 6]
```diff
@@ -144,16 +145,28 @@ export class FastConfirmationRule implements IFastConfirmationRule {
     }
   }
 
-  private updateFastConfirmationMetrics(ctx: FastConfirmationContext, result: FastConfirmationResult): void {
+  private updateFastConfirmationMetrics(
+    ctx: FastConfirmationContext,
+    result: Pick<FastConfirmationRunResult, "confirmedRoot" | "didReset" | "didReorg" | "didFallback" | "didRestart">
```
Comment (nflaig) [inline packages/fork-choice/src/forkChoice/fastConfirmation/fastConfirmationRule.ts]: the pick is kinda ugly but if you are fine with it @nazarhussain I am ok with it too

**Review by nflaig — COMMENTED** [review]
LGTM, thanks @nazarhussain for picking this up

File: dashboards/lodestar_fork_choice.json [hunk 7]
```diff
@@ -1692,13 +1692,13 @@
           "fullMetaSearch": false,
           "includeNullMetadata": true,
           "instant": false,
-          "legendFormat": "Reset Total",
+          "legendFormat": "Resets",
```
Comment (nflaig) [inline dashboards/lodestar_fork_choice.json:1695]: btw it's still "reset total" but not since startup of the node (as before) but now per selected rate interval, either is fine, I don't think the "total" is necessarily wrong or inaccurate here though

Comment (nazarhussain) [reply dashboards/lodestar_fork_choice.json:1695]: `Resets Total` refers to accumulated value, but when using `rate` it's much aligned to just have `Resets`

**Review by nflaig — COMMENTED** [review]
LGTM, did a final pass

**Review by wemeetagain — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.44.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.44.0) :tada:
